pbft/pbft.cpp

```diff
@@ -994,6 +1007,533 @@ pbft::get_high_water_mark()
+    {
+        // --- add the sequence/hash to a set
+        auto prepared_proof = viewchange_msg.prepared_proofs(j);
+
+        pbft_msg msg;
+        if (bzn_envelope envelope;
+            !envelope.ParseFromString(prepared_proof.pre_prepare())
+            || !this->is_peer(envelope.sender())
+            || !msg.ParseFromString(envelope.pbft()))
+        {
+            return false;
+        }
+        sequence_request_pairs.insert(std::make_pair(msg.sequence(), msg.request_hash()));
+    }
+    return true;
+}
+
+/**
+ * A valid newview message must contain a complete set of pre prepares, that is
+ * there must be one pre prepare for each sequence.
+ * TODO: this method needs a better name
+ * @param newview_msg
+ * @return true if there are no missing pre-prepares based on sequence numbers of each pre-prepare, false otherwise.
+ */
+bool
+pbft::pre_prepares_contiguous(const pbft_msg& newview_msg) const
+{
+    uint64_t minimum_sequence{UINT64_MAX};
+    uint64_t maximum_sequence{0};
+    for (int i{0}; i < newview_msg.pre_prepare_messages_size(); ++i)
+    {
+        const bzn_envelope pre_prepare_envelope{newview_msg.pre_prepare_messages(i)};
+        pbft_msg pre_prepare;
+        pre_prepare.ParseFromString(pre_prepare_envelope.pbft());
+
+        minimum_sequence = pre_prepare.sequence() < minimum_sequence ? pre_prepare.sequence() : minimum_sequence;
+        maximum_sequence = pre_prepare.sequence() > maximum_sequence ? pre_prepare.sequence() : maximum_sequence;
+    }
+    return (maximum_sequence-minimum_sequence) == uint64_t(newview_msg.pre_prepare_messages_size());
+}
+
+bool
+pbft::is_valid_newview_message(const pbft_msg& msg, const bzn_envelope& /*original_msg*/) const
+{
+    // - does it contain 2f+1 viewchange messages
+    if (static_cast<size_t>(msg.viewchange_messages_size()) < 2 * this->max_faulty_nodes() + 1)
+    {
+        return false;
+    }
+
+    if (!this->pre_prepares_contiguous(msg))
+    {
+        LOG (error) << "is_valid_newview_message - NEWVIEW is missing pre prepare messages";
+        return false;
+    }
+
+    pbft_msg viewchange_msg;
+    std::set<std::string> viewchange_senders;
+    for (int i{0};i < msg.viewchange_messages_size();++i)
+    {
+        bzn_envelope original_msg;
+        // - are each of those viewchange messages valid?
+        if (!original_msg.ParseFromString(viewchange_msg.viewchange_messages(i)) || !viewchange_msg.ParseFromString(original_msg.pbft()) || !this->is_valid_viewchange_message(viewchange_msg, original_msg))
+        {
+            return false;
+        }
+
+        viewchange_senders.insert(original_msg.sender());
+        std::set<std::pair<uint64_t, std::string>> sequence_request_pairs;
+
+        if (!this->get_sequences_and_request_hashes_from_proofs(viewchange_msg, sequence_request_pairs))
+        {
+            return false;
+        }
+
+        // each *pair* must match a pre-prepare in the the newview message...
```

Comment:
My fault for forgetting to tell you this detail, but:
operations that come after the checkpoint which the new-view bases off of should be included.
So for example, if node A sends a view-change with checkpoint 100 and operations 102 and 105, but operation B sends a view change with checkpoint 200, then the new-view should use checkpoint 200, and should not have new preprepares for operations 102 and 105, because they're already included in checkpoint 200, and checkpoint 200 is proven valid.

Reply:
In build new view, before collating pre prepare proofs, we ignore pre prepares with a sequence number lower than the highest checkpoint sequence number in the view change messages received from all of the nodes.